group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,891 @@
+
+        if (currentAssignmentSize < maxAssignmentSize)
+            return true;
+
+        for (Uuid topicId: assignedTopicIds) {
+            if (canTopicParticipateInReassignment(topicId))
+                return true;
+        }
+        return false;
+    }
+
+    /**
+     * Determine if the current assignment is a balanced one.
+     *
+     * @return true if the given assignment is balanced; false otherwise.
+     */
+    private boolean isBalanced() {
+        int min = assignmentManager.targetAssignmentSize(sortedMembersByAssignmentSize.first());
+        int max = assignmentManager.targetAssignmentSize(sortedMembersByAssignmentSize.last());
+
+        // If minimum and maximum numbers of partitions assigned to consumers differ by at most one return true.
+        if (min >= max - 1)
+            return true;
+
+        // For each member that does not have all the topic partitions it can get make sure none of the
+        // topic partitions it could but did not get cannot be moved to it (because that would break the balance).
+        // Members with the least assignment sizes are checked first to see if they can receive any more partitions.
+        for (String member: sortedMembersByAssignmentSize) {
+            int memberPartitionCount = assignmentManager.targetAssignmentSize(member);
+
+            // Skip if this member already has all the topic partitions it can get.
+            List<Uuid> allSubscribedTopics = new ArrayList<>(members.get(member).subscribedTopicIds());
+            int maxAssignmentSize = assignmentManager.maxAssignmentSize(member);
+
+            if (memberPartitionCount == maxAssignmentSize)
+                continue;
+
+            // Otherwise make sure it cannot get any more partitions.
+            for (Uuid topicId: allSubscribedTopics) {
+                Set<Integer> assignedPartitions = targetAssignment.get(member).targetPartitions().get(topicId);
+                for (int i = 0; i < subscribedTopicDescriber.numPartitions(topicId); i++) {
+                    TopicIdPartition topicIdPartition = new TopicIdPartition(topicId, i);
+                    if (assignedPartitions == null || !assignedPartitions.contains(i)) {
+                        String otherMember = partitionOwnerInTargetAssignment.get(topicIdPartition);
+                        int otherMemberPartitionCount = assignmentManager.targetAssignmentSize(otherMember);
+                        if (memberPartitionCount + 1 < otherMemberPartitionCount) {
+                            LOG.debug("{} can be moved from member {} to member {} for a more balanced assignment.",
+                                topicIdPartition, otherMember, member);
+                            return false;
+                        }
+                    }
+                }
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Balance the current assignment using the data structures created in the assignPartitions(...) method above.
```

Comment:
I would put a link to the actual method. I don't see any method called `assignPartitions`

Reply:
you're right, let me update the java docs, thanks